### PR TITLE
[StatefulTransform] Extract `scf.for` unrolling logic

### PR DESCRIPTION
**Pull request by jtuyls**
Reduces the scope of the `AMDAIEObjectFifoStatefulTransform` pass by extracting the unrolling logic and moving it up into `AMDAIE`.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 1]
```diff
@@ -79,10 +79,27 @@ AIE::BDDimLayoutArrayAttr convertSizeStrideToBDDimLayoutArrayAttr(
...
+           << "both source and target on main memory not supported";
+  } else if (sourceMemSpace == 0) {
+    depth = targetDepth;
+  } else if (targetMemSpace == 0) {
+    depth = sourceDepth;
+  } else {
+    if (sourceDepth != targetDepth)
+      return dmaOp.emitOpError() << "unsupported sourceDepth != targetDepth";
+    depth = sourceDepth;
+  }
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:101]: > L3 -> L2 : this condition is not checked.
L2 -> L3 : this condition is not checked.

These are handled here: https://github.com/nod-ai/iree-amd-aie/pull/666/files#diff-888fdafbbd093dc98e4fae55a26e7fa56065fc7506fedc0107b8a897c19f816aR93

> So, sourceDepth != targetDepth is unsupported for transactions between L1/L2. Is that correct?

It's not supported in `aie.objectfifo`, so I am throwing an error here.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+
+#define DEBUG_TYPE "iree-amdaie-core-loop-unroll"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Unroll the scf.for loops inside the core operations based on the depths of
+/// the acquired objFifos.
+LogicalResult coreLoopUnroll(RewriterBase &rewriter, AMDAIE::CoreOp coreOp) {
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp:23]: This pass does a different unrolling from the `DistributeCoresAndObjectFifos ` and others as it checks for synchronzation ops + calculates a different unroll factor.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+
+#define DEBUG_TYPE "iree-amdaie-assign-logical-objectfifo-depth"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Depth is hardcoded to '2' for now, meaning double buffering is enabled by
+/// default.
+static const unsigned kDefaultLogicalObjFifoDepth = 2;
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp:17]: Yeah, makes sense, will add that,

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+void AMDAIEAssignLogicalObjectFifoDepthPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(parentOp->getContext());
+  // Only assign for L2 and L1, not for L3 (main memory).
+  WalkResult res = parentOp->walk(
+      [&](AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo) {
+        uint8_t memSpace = logicalObjectFifo.getMemorySpaceAsUInt();
+        if (memSpace != 1 && memSpace != 2) return WalkResult::advance();
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp:38]: This should be more future proof (safe) I think, i.e. if there would be a memSpace 3, 4 etc in the future.

**Comment by newling** [comment]
Are the 2 new passes here both doing something stateful-transform pass was doing? Is it possible to make 2 PRs for the 2 new passes?

**Comment by jtuyls** [comment]
> Are the 2 new passes here both doing something stateful-transform pass was doing? Is it possible to make 2 PRs for the 2 new passes?

I extracted the buffer depth assignment into https://github.com/nod-ai/iree-amd-aie/pull/672.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp]: Based on @newling's request I put this in a new PR: https://github.com/nod-ai/iree-amd-aie/pull/672

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignLogicalObjectFifoDepth.cpp]: This was moved and changed in https://github.com/nod-ai/iree-amd-aie/pull/672. Could you have another look?

**Comment by jtuyls** [comment]
@Abhishek-Varma I rebased this on top of https://github.com/nod-ai/iree-amd-aie/pull/672. Could you have another look?

**Review by newling — APPROVED** [review]
LGTM, just 3 small suggestions

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                      stridedOp.getTarget().getDefiningOp())
+                : dyn_cast<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+                      stridedOp.getSource().getDefiningOp());
+        depths.insert(
+            cast<LogicalObjectFifoType>(logicalObjFifo.getType()).getDepth());
+      }
+    }
+    int unrollFactor =
+        std::accumulate(depths.begin(), depths.end(), 1, std::lcm<int, int>);
+    if (found && failed(mlir::loopUnrollByFactor(forOp, unrollFactor))) {
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp]: Yeah, thanks for the suggestion. I just copied the code, but this was done a bit weirdly indeed.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  WalkResult res = coreOp.walk([&](scf::ForOp forOp) {
+    bool found = false;
+    llvm::SmallDenseSet<unsigned> depths;
+    for (auto acqOp :
+         forOp.getBody()->getOps<AMDAIE::LogicalObjectFifoAcquire>()) {
+      if (acqOp.getOperation()->getParentOp() == forOp) {
+        found = true;
+        auto stridedOp = dyn_cast<DoublyStridedCopyOpInterface>(
+            acqOp.getDma().getDefiningOp());
+        if (!stridedOp) return WalkResult::interrupt();
```
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECoreLoopUnroll.cpp]: Done

**Pull request merged**